packages/app-builder/src/routes/_builder+/scenarios+/$scenarioId+/i+/$iterationId+/sanction.tsx

```
@@ -150,31 +150,32 @@ export async function action({ request, params }: ActionFunctionArgs) {
     toastSessionService: { getSession, commitSession },
   } = serverServices;
 
-  const session = await getSession(request);
-  const formData = await request.formData();
-  const iterationId = fromParams(params, 'iterationId');
-
-  const submission = parseWithZod(formData, { schema: editSanctionFormSchema });
+  const [session, formData, { scenarioIterationSanctionRepository }] =
+    await Promise.all([
+      getSession(request),
+      request.formData(),
+      authService.isAuthenticated(request, {
+        failureRedirect: getRoute('/sign-in'),
+      }),
+    ]);
 
-  if (submission.status !== 'success') {
-    return json(submission.reply());
-  }
-
-  const { scenarioIterationSanctionRepository } =
-    await authService.isAuthenticated(request, {
-      failureRedirect: getRoute('/sign-in'),
-    });
+  const iterationId = fromParams(params, 'iterationId');
+  const formDataDecoded = formDataToObject(formData, {
+    arrays: ['datasets'],
+  });
 
   try {
+    const data = editSanctionFormSchema.parse(formDataDecoded);
```

why does it go in the `try` now ? Couldn't you use `safeParse` with and return an object `{ success: false, errors: errors[] }`  or `{ success: true, data }` ?

---

I put it in the try block because `parse` throw if it cannot parse, that way I don't have to right the exact same code present in the catch block just before the try-catch ;)

---

Ok, so the failed validation is an unknown error?

---

And we do not need to return the error since the message shown with the Toast do not include it